jerry-core/ecma/builtin-objects/ecma-builtin-helpers.c

```diff
@@ -450,6 +449,71 @@ ecma_builtin_helper_string_index_normalize (ecma_number_t index, /**< index */
   return norm_index;
 } /* ecma_builtin_helper_string_index_normalize */
 
+/**
+ * Helper function for finding lastindex of a search string
+ *
+ * This function clamps the given index to the [0, length] range.
+ * If the index is negative, 0 value is used.
+ * If the index is greater than the length of the string, the normalized index will be the length of the string.
+ *
+ * See also:
+ *          ECMA-262 v6, 21.1.3.9
+ *
+ * Used by:
+ *         - The ecma_builtin_helper_string_prototype_object_index_of helper routine.
+ *         - The ecma_builtin_string_prototype_object_replace_match helper routine.
+ *
+ * @return uint32_t - whether there is a match for the search string
+ */
+static uint32_t
+ecma_builtin_helper_string_find_last_index (ecma_string_t *original_str_p, /**< original string */
+                                            ecma_string_t *search_str_p, /**< search string */
+                                            uint32_t position) /**< start_position */
+{
+  if (ecma_string_is_empty (search_str_p))
+  {
+    return position;
+  }
+
+  uint32_t original_length = ecma_string_get_length (original_str_p);
+
+  ECMA_STRING_TO_UTF8_STRING (search_str_p, search_str_utf8_p, search_str_size);
+  ECMA_STRING_TO_UTF8_STRING (original_str_p, original_str_utf8_p, original_str_size);
+
+  const lit_utf8_byte_t *end_p = original_str_utf8_p + original_str_size;
+  const lit_utf8_byte_t  *current_p = end_p;
+
+  for (ecma_number_t i = original_length; i > position; i--)
+  {
+    lit_utf8_decr (&end_p);
+  }
+
+  uint32_t ret_value = UINT32_MAX;
+
+  while (true)
+  {
+    if (end_p + search_str_size <= current_p
```

Comment:
This first part of the condition could be removed, if you further adjust the starting search position before the main while loop.

```c
while (current_p + search_str_size > end_p)
{
  lit_utf8_decr (&current_p)
  position--;
}
```